### PR TITLE
Implement CMake-Buildsystem for Beckhoff ADS.

### PR DESCRIPTION
**Pull request by miq**

### CODE REVIEW

File: example/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+add_executable(example.bin ${SOURCES})
+
+target_link_libraries(example.bin ads pthread)
+
+target_compile_features(example.bin PRIVATE
+  cxx_std_11
+)
+
+target_compile_options(example.bin PRIVATE
```
Comment (pbruenn) [inline example/CMakeLists.txt:17]: I am not a CMake expert, could we put this in the toplevel CMakeList? As all our subdirs _should_ use the same compile features and options.

**Review by pbruenn — CHANGES_REQUESTED** [review]
This is definitely interesting. Some years ago I tried to get rid of the visual studio solutions with help of CMake. I think I used the wrong options to make CMake work for visual studio and plain CMake on Linux, MacOS and FreeBSD. I didn't found time to test this on all of our targets but if we can replace both Makefiles and visual studio projects with a clean set of CMakeLists.txt I will merge it (and drop the Makefiles, too!).

**Comment by miq** [comment]
I moved the features, options and definitions to the top-level CMakeLists.txt. It works at least for our target platform (GCC on Debian Linux).

**Comment by pbruenn** [comment]
Thanks for the update. I looked deeper into CMake, too and found Meson...
The problem with CMake is I'm unable to get rid of these compiler specific debug level and library includes. Meson does fix this very well. I'm aware CMake seems to be the defacto standard for C++ builds, like MS exchange seems to mailservers. But my hope is that people will adapt to the much saner Meson.
Iam neither an expert on CMake nor Meson, but just from the look on the configuration files to support all of AdsLibs platforms Meson does look much nicer to me. We have some limited experience with CMake within Beckhoff, but these just confirm my fear that CMake setups end up in a mess of conditionals.
So I'm pretty sorry for you to say, I will not integrate CMake, but instead work on Meson support. Unless you convince me with a pretty example which build all of our platforms nicely. This is my current state:
```
project('AdsLib', 'cpp',
  version : '0.1',
  default_options : [
    'cpp_std=c++11',
    'warning_level=3',
  ]
)

common_files = files([
  'AdsLib/AdsDef.cpp',
  'AdsLib/AdsDevice.cpp',
  'AdsLib/AdsFile.cpp',
  'AdsLib/Log.cpp',
  'AdsLib/Sockets.cpp',
  'AdsLib/Frame.cpp',
])

router_files = files([
  'AdsLib/standalone/AdsLib.cpp',
  'AdsLib/standalone/AmsConnection.cpp',
  'AdsLib/standalone/AmsNetId.cpp',
  'AdsLib/standalone/AmsPort.cpp',
  'AdsLib/standalone/AmsRouter.cpp',
  'AdsLib/standalone/NotificationDispatcher.cpp',
])

inc = include_directories([
  'AdsLib',
  'tools',
])

libs = [
  dependency('threads'),
]

libs += meson.get_compiler('cpp').find_library('ws2_32', required: false)


adslib = static_library('AdsLib',
  [common_files, router_files],
  include_directories: inc,
)

adslibtest = executable('AdsLibTest',
  'AdsLibTest/main.cpp',
  include_directories: inc,
  dependencies: libs,
  link_with: adslib)

adslibooitest = executable('AdsLibOOITest',
  'AdsLibOOITest/main.cpp',
  include_directories: inc,
  dependencies: libs,
  link_with: adslib)
```

**Comment by pbruenn** [comment]
This is the full meson version:
https://github.com/Beckhoff/ADS/tree/patrickbr/meson

**Comment by miq** [comment]
Great that you implemented a new build system for the library. I am sure many users will appreciate it.
For us the great tooling around CMake for CLion, QtCreator and Visual Studio makes the difference. If you do not want to merge my changes for whatever reason, feel free to do so.
Then we will maintain our fork for the people who prefer CMake.

Thank you anyway!

**Comment by pbruenn** [comment]
@miq I really appreciate your work. I think the only argument I have against merging is people might complain if something breaks or isn't working. But in that case I can always link your git account ;-). So just tell me, when you think your changes are ready to merge and I will hit the button. Do you want the commits squashed?

**Comment by miq** [comment]
From my side the changes are ready now. I am fine with delegating people with problems to me, since we have only tested it on linux machines.

Regarding squashing: I really do not care, merge as you like it best!

Thanks.

**Pull request merged**